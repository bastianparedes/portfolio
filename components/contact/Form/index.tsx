import React, { useState } from 'react';
import styles from './styles.module.scss';
import Modal from '../Modal';

interface typeElements extends HTMLFormControlsCollection {
  userName: { value: string };
  email: { value: string };
  subject: { value: string };
  message: { value: string };
}

interface typeTarget {
  elements: typeElements;
}

const Form = (): JSX.Element => {
  const [modalVisible, setModalVisible] = useState(false);
  const [submitSuccessful, setSubmitSuccessful] = useState(true);
  const [modalText, setModalText] = useState('');

  const onSubmit = (event: React.FormEvent): void => {
    event.preventDefault();
    const target = event.target as typeof event.target & typeTarget;
    const elements = target.elements;
    const userName = elements.userName.value;
    const email = elements.email.value;
    const subject = elements.subject.value;
    const message = elements.message.value;

    fetch('/api/contact', {
      method: 'POST',
      body: JSON.stringify({
        subject: `${userName} quiere contactarte.`,
        text: [
          `Nombre: ${userName}`,
          `Subject: ${subject}`,
          `E-mail: ${email}`,
          `Message: ${message}`
        ].join('\n')
      })
    })
      .then((response) => {
        if (response.ok) {
          setSubmitSuccessful(true);
          setModalText('Mensaje enviado correctamente');
        } else {
          setSubmitSuccessful(false);
          setModalText(
            'Ha ocurrido un error inesperado, por favor inténtelo más tarde'
          );
        }
      })
      .catch(() => {
        setSubmitSuccessful(false);
        setModalText(
          'Ha ocurrido un error inesperado, por favor inténtelo más tarde'
        );
      })
      .finally(() => {
        setModalVisible(true);
      });
  };

  return (
    <>
      <form
        data-testid="contact-form"
        onSubmit={onSubmit}
        className={styles.form}
      >
        <input name="userName" type="text" placeholder="Nombre" />
        <input name="email" type="email" placeholder="Correo electrónico" />
        <input name="subject" type="text" placeholder="Asunto" />
        <textarea
          name="message"
          placeholder="Mensaje"
          className={styles.textarea}
        />
        <button type="submit" value="Enviar">
          Enviar
        </button>
      </form>
      {modalVisible && (
        <Modal
          success={submitSuccessful}
          text={modalText}
          setModalVisible={setModalVisible}
        />
      )}
    </>
  );
};

export default Form;
