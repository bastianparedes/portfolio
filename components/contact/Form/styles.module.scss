.formContainer {
  display: flex;
  flex-direction: column;
  justify-content: center;
}

.form {
  display: grid;
  gap: 1rem;
}

.textarea {
  resize: none;
  height: 100px;
}
