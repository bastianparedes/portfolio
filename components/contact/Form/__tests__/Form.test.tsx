import React from 'react';

import { fireEvent, render } from '@testing-library/react';
import { act } from 'react-dom/test-utils';

import Form from '..';

jest.mock(
  '../../ResultModal',
  () =>
    ({ setModalVisible }: { setModalVisible: (boolean: boolean) => void }) => {
      const handleOnClose = (): void => {
        setModalVisible(false);
      };
      return <button onClick={handleOnClose}>X</button>;
    }
);
jest.mock('bastianparedes/components');

describe('<Form />', () => {
  it('should render', () => {
    const { container } = render(<Form />);
    expect(container).toMatchSnapshot();
  });

  it('should call api/contact and success with response.ok true', async () => {
    (global.fetch as jest.Mock) = jest.fn(
      async () => await Promise.resolve({ ok: true })
    );
    const { getByText } = render(<Form />);

    await act(async () => {
      fireEvent.click(getByText('Enviar'));
    });

    expect(fetch).toBeCalledTimes(1);
  });

  it('should call api/contact and success with response.ok false', async () => {
    (global.fetch as jest.Mock) = jest.fn(
      async () => await Promise.resolve({ ok: false })
    );
    const { getByText } = render(<Form />);
    await act(async () => {
      fireEvent.click(getByText('Enviar'));
    });

    expect(fetch).toBeCalledTimes(1);
  });

  it('should call api/contact and fail', async () => {
    global.fetch = jest.fn(async () => await Promise.reject(new Error()));
    const { getByText } = render(<Form />);
    await act(async () => {
      fireEvent.click(getByText('Enviar'));
    });

    expect(fetch).toBeCalledTimes(1);
  });

  it('should close modal after it is opened', async () => {
    (global.fetch as jest.Mock) = jest.fn(
      async () => await Promise.resolve({ ok: true })
    );
    const { getByText } = render(<Form />);
    await act(async () => {
      fireEvent.click(getByText('Enviar'));
    });

    const buttonCloserModal = getByText('X');
    expect(buttonCloserModal).toBeInTheDocument();
    await act(async () => {
      fireEvent.click(buttonCloserModal);
    });
    expect(buttonCloserModal).not.toBeInTheDocument();
  });
});
