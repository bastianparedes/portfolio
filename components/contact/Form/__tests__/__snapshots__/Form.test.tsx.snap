// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Form /> should render 1`] = `
<div>
  <form
    class="form"
    data-testid="contact-form"
  >
    <input
      name="userName"
      placeholder="Nombre"
      type="text"
    />
    <input
      name="email"
      placeholder="Correo electrÃ³nico"
      type="email"
    />
    <input
      name="subject"
      placeholder="Asunto"
      type="text"
    />
    <textarea
      class="textarea"
      name="message"
      placeholder="Mensaje"
    />
    <button
      type="submit"
      value="Enviar"
    >
      Enviar
    </button>
  </form>
</div>
`;
